test/unit/sharding_key_test.lua

```diff
@@ -0,0 +1,87 @@
+local t = require('luatest')
+local sharding_key_module = require('crud.common.sharding_key')
+local utils = require('crud.common.utils')
+
+local g = t.group('sharding_key')
+
+g.before_each(function()
+    local sharding_key_format = {
+        {name = 'space_name', type = 'string', is_nullable = false},
+        {name = 'sharding_key', type = 'array', is_nullable = false}
+    }
+    -- Create a space _ddl_sharding_key with a tuple that
+    -- contains a space name and it's sharding key.
+    if type(box.cfg) ~= 'table' then
+        box.cfg{}
+    end
+    box.schema.space.create('_ddl_sharding_key', {
+        format = sharding_key_format,
+        if_not_exists = true
+    })
+    box.space._ddl_sharding_key:create_index('pk', {if_not_exists = true})
+end)
+
+g.after_each(function()
+    -- Cleanup.
+    box.space._ddl_sharding_key:truncate()
+    box.space._ddl_sharding_key:drop()
+end)
+
+g.test_as_index_object_positive = function()
+    local space_name = 'as_index_object'
+    local space_format = {
+        {name = 'id', type = 'unsigned'},
+        {name = 'name', type = 'string'},
+        {name = 'age', type = 'unsigned'},
+    }
+    local sharding_key_def = {'name', 'age'}
+
+    local index_obj, err = sharding_key_module.as_index_object(space_format,
+                                                               space_name,
+                                                               sharding_key_def)
+    t.assert_equals(err, nil)
+    t.assert_equals(index_obj, {
+        parts = {
+            {fieldno = 2},
+            {fieldno = 3},
+        }
+    })
+end
+
+g.test_as_index_object_negative = function()
+    local space_name = 'as_index_object'
+    local space_format = {
+        {name = 'id', type = 'unsigned'},
+        {name = 'name', type = 'string'},
+        {name = 'age', type = 'unsigned'},
+    }
+    local sharding_key_def = {'dude', 'age'}
+
+    local index_obj, err = sharding_key_module.as_index_object(space_format,
+                                                               space_name,
+                                                               sharding_key_def)
+    t.assert_str_contains(err.err,
+        'No such field (dude) in a space format (as_index_object)')
+    t.assert_equals(index_obj, nil)
+end
+
+g.test_get_format_fieldno_map = function()
+    local space_format = {
+        {name = 'id', type = 'unsigned'},
+        {name = 'name', type = 'string'},
+        {name = 'age', type = 'unsigned'},
+    }
+
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    t.assert_equals(fieldno_map, {age = 3, id = 1, name = 2})
+end
+
+g.test_fetch_on_storage = function()
+    local space_name = 'fetch_on_storage'
+    local sharding_key_def = {'name', 'age'}
+    box.space._ddl_sharding_key:upsert(
+        {space_name, sharding_key_def}, {{'=', 1, space_name}})
```

Comment:
Do you attempt to update the primary key (`{'space_name'}`) if the tuple exists? That's looks wrong (and I guess it will not work). I guess you meant to update the second field (`'sharding_key'`).

Reply:
If we'll consider https://github.com/tarantool/crud/pull/181/commits/efa753de399f725359e9e1efc2e01adc693ea0d4#r713208010 it looks meaningful to just `:insert()` here.

Reply:
replaced `upsert` with `insert`:

```lua
--- a/test/unit/sharding_key_test.lua
+++ b/test/unit/sharding_key_test.lua
@@ -79,8 +79,7 @@ end
 g.test_fetch_on_storage = function()
     local space_name = 'fetch_on_storage'
     local sharding_key_def = {'name', 'age'}
-    box.space._ddl_sharding_key:upsert(
-        {space_name, sharding_key_def}, {{'=', 1, space_name}})
+    box.space._ddl_sharding_key:insert({space_name, sharding_key_def})
 
     local map_with_keys = sharding_key_module.fetch_on_storage()
     t.assert_equals(map_with_keys, {[space_name] = sharding_key_def})
```